src/common/Throttle.cc

```diff
@@ -729,27 +750,71 @@ void TokenBucketThrottle::set_max(uint64_t m) {
 }
 
 void TokenBucketThrottle::set_average(uint64_t avg) {
-  m_avg = avg;
+  {
+    std::lock_guard<Mutex> lock(m_lock);
+    m_avg = avg;
+
+    if (0 == avg) {
+      // this means No Limit
+      m_tick = 1000;
+    } else {
+      // calc the tick(ms), don't less than the minimum.
+      m_tick = 1000 / avg;
+      if (m_tick < m_tick_min) {
+        m_tick = m_tick_min;
+      }
+
+      // tokens to put into the bucket every tick.
+      m_tokens_per_tick = avg * m_tick / 1000;
+
+      // this is for the number(avg) can not be divisible.
+      m_ticks_per_second = 1000 / m_tick;
+      m_divide_left = avg % m_ticks_per_second;
+
+      // zero means burst IO is disabled.
+      if (0 == m_throttle.max) {
+        m_throttle.set_max(m_tokens_per_tick);
+      } else {
+        m_throttle.remain = m_throttle.max;
```

Comment:
why we need to set remain directly out of Bucket? Actually max and remain should be private, and we should access them by Bucket.get(), Bucket.put() and Bucket.set_max().  I am not sure why we need other operations on a bucket.